### ISSUE TITLE
Problem with relocations

### ISSUE THREAD

**Issue by sdasgup3** [revng/revamb#7]
After building the project on orchestra, I tried building hello world program as depicted [here](https://github.com/revng/revamb#example-run), 
```
$$ ./root/usr/x86_64-pc-linux-gnu/armv7a-hardfloat-linux-uclibceabi/gcc-bin/4.9.3/armv7a-hardfloat-linux-uclibceabi-gcc -static hello.c -o hello.arm
$$ bash -x ./root/bin/translate -s  hello.arm
+++ dirname ./root/bin/translate
++ cd ./root/bin
++ pwd
+ SCRIPT_PATH=/home/sdasgup3/Github/orchestra/root/bin
+ INPUT=
+ OPTIMIZE=0
+ SKIP=0
+ SUPPORT_CONFIG=normal
+ set -e
+ [[ 2 > 0 ]]
+ key=-s
+ case $key in
+ SKIP=1
+ shift
+ [[ 1 > 0 ]]
+ key=hello.arm
+ case $key in
+ '[' -z '' ']'
+ INPUT=hello.arm
+ shift
+ [[ 0 > 0 ]]
+ LL=hello.arm.ll
+ LINKED_LL=hello.arm.linked.ll
+ REVAMB_LOG=hello.arm.ll.log
+ LL_OPT=hello.arm.opt.ll
+ CSV=hello.arm.ll.li.csv
+ OBJ=hello.arm.ll.o
+ export PATH=/home/sdasgup3/Github/orchestra/root/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
+ PATH=/home/sdasgup3/Github/orchestra/root/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

+ CC=cc
+ LINK=llvm-link
+ LLC=llc
+ OPT=opt
+ REVAMB=revamb
+ TOOPT=li-csv-to-ld-options
++ python -c '
from binascii import hexlify
import sys
file = open(sys.argv[1], "rb")
file.seek(5)
result = hexlify(file.read(1))
file.seek(17)
result += hexlify(file.read(2))
file.close()
print(result.decode("utf-8"))' hello.arm
+ ARCHID=010028
+ case "$ARCHID" in
+ ARCH=arm
+ SUPPORT_NAME=support-arm-normal.ll
+ SUPPORT_PATH=/home/sdasgup3/Github/orchestra/root/bin/../share/revamb/support-arm-normal.ll
+ '[' '!' -e /home/sdasgup3/Github/orchestra/root/bin/../share/revamb/support-arm-normal.ll ']'
+ '[' 1 -eq 0 ']'
+ llvm-link hello.arm.ll /home/sdasgup3/Github/orchestra/root/bin/../share/revamb/support-arm-normal.ll -o hello.arm.linked.ll -S
+ OUTPUT=hello.arm.translated
+ '[' 0 -eq 0 ']'
+ llc -O0 -filetype=obj hello.arm.linked.ll -o hello.arm.ll.o
++ li-csv-to-ld-options hello.arm.ll.li.csv
+ cc -fuse-ld=bfd -Wl,--section-start=.elfheaderhelper=0xffff -Wl,-Ttext-segment=0x30000 -Wl,-z,max-page-size=4096 -Wl,--section-start=.o_rx_0x10000=0x10000 -Wl,--section-start=.o_rw_0x2d08c=0x2d08c hello.arm.ll.o -lz -lm -lrt -o hello.arm.translated -fno-pie
/usr/bin/ld.bfd: hello.arm.ll.o: relocation R_X86_64_32S against `.bss' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld.bfd: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status

``` 
I am suspecting this against building uclibc statically! Can you please throw some light to it.

**Comment by aleclearmind** [comment]
Hi, sorry for the delayed answer.

What version of GCC are you using?
Can you please add in `translate` script `-no-pie` on the final invocation to $CC?

**Comment by sdasgup3** [comment]
Thanks for the reply. 
$ gcc --version
``gcc (Ubuntu 6.3.0-12ubuntu2) 6.3.0 20170406``
I tried no-pie and able to generate the executable which segfaults on execution.

**Comment by aleclearmind** [comment]
Can you upload the program you're translating so I can give it a try?
What's the result of running revamb's the test suite (`make test`)?

**Comment by sdasgup3** [comment]
Thanks. This is the program:
```C
#include <stdio.h>

int main(int argc, char *argv[]) {
  printf("Hello, world!\n");
  return 0;
}
```
and here is the summary of ctests
```
50% tests passed, 78 tests failed out of 156

Label Time Summary:
analysis                      =  15.65 sec (35 tests)
arm                           = 172.52 sec (50 tests)
calc                          =  54.94 sec (45 tests)
cfg                           =   1.36 sec (7 tests)
check-with-native             =   0.07 sec (21 tests)
check-with-qemu               =   0.42 sec (21 tests)
check-with-reference          =   4.64 sec (21 tests)
compile-translated            =  31.61 sec (15 tests)
default                       =   3.29 sec (52 tests)
extract-info                  =   2.14 sec (7 tests)
floating_point                =  51.72 sec (19 tests)
function_call                 =  51.02 sec (19 tests)
functionsboundaries           =   1.43 sec (7 tests)
global                        =  49.99 sec (19 tests)
literal                       =   0.58 sec (13 tests)
memset                        =   0.63 sec (3 tests)
memset-arm                    =   2.04 sec (5 tests)
mips                          =  46.95 sec (41 tests)
multiplication                =   0.54 sec (13 tests)
noreturn                      =   1.85 sec (7 tests)
run-qemu-test                 =   1.22 sec (21 tests)
run-test-native               =   0.12 sec (7 tests)
run-translated-test           =   3.16 sec (21 tests)
runtime                       = 257.36 sec (121 tests)
sum                           =   0.57 sec (13 tests)
switch-addls                  =   0.62 sec (3 tests)
switch-addls-arm              =   2.00 sec (5 tests)
switch-disjoint-ranges        =   1.11 sec (3 tests)
switch-disjoint-ranges-arm    =   2.62 sec (5 tests)
switch-jump-table             =   1.75 sec (6 tests)
switch-jump-table-mips        =   2.28 sec (5 tests)
switch-jump-table-x86_64      =   3.11 sec (5 tests)
switch-ldrls                  =   0.22 sec (3 tests)
switch-ldrls-arm              =   1.79 sec (5 tests)
syscall                       =  49.70 sec (19 tests)
translate                     = 229.64 sec (22 tests)
try-catch-ehframe             =   0.32 sec (3 tests)
try-catch-ehframe-x86_64      =   1.81 sec (5 tests)
x86_64                        =  42.42 sec (44 tests)

Total Test time (real) =  38.63 sec

The following tests FAILED:
	  9 - compile-translated-calc-arm (Failed)
	 10 - run-translated-test-calc-literal-arm (Failed)
	 11 - check-with-native-calc-literal-arm (Failed)
	 13 - check-with-qemu-calc-literal-arm (Failed)
	 14 - run-translated-test-calc-sum-arm (Failed)
	 15 - check-with-native-calc-sum-arm (Failed)
	 17 - check-with-qemu-calc-sum-arm (Failed)
	 18 - run-translated-test-calc-multiplication-arm (Failed)
	 19 - check-with-native-calc-multiplication-arm (Failed)
	 21 - check-with-qemu-calc-multiplication-arm (Failed)
	 23 - compile-translated-function_call-arm (Failed)
	 24 - run-translated-test-function_call-default-arm (Failed)
	 25 - check-with-native-function_call-default-arm (Failed)
	 27 - check-with-qemu-function_call-default-arm (Failed)
	 29 - compile-translated-floating_point-arm (Failed)
	 30 - run-translated-test-floating_point-default-arm (Failed)
	 31 - check-with-native-floating_point-default-arm (Failed)
	 33 - check-with-qemu-floating_point-default-arm (Failed)
	 35 - compile-translated-syscall-arm (Failed)
	 36 - run-translated-test-syscall-default-arm (Failed)
	 37 - check-with-native-syscall-default-arm (Failed)
	 39 - check-with-qemu-syscall-default-arm (Failed)
	 41 - compile-translated-global-arm (Failed)
	 42 - run-translated-test-global-default-arm (Failed)
	 43 - check-with-native-global-default-arm (Failed)
	 45 - check-with-qemu-global-default-arm (Failed)
	 47 - compile-translated-calc-mips (Failed)
	 48 - run-translated-test-calc-literal-mips (Failed)
	 49 - check-with-native-calc-literal-mips (Failed)
	 51 - check-with-qemu-calc-literal-mips (Failed)
	 52 - run-translated-test-calc-sum-mips (Failed)
	 53 - check-with-native-calc-sum-mips (Failed)
	 55 - check-with-qemu-calc-sum-mips (Failed)
	 56 - run-translated-test-calc-multiplication-mips (Failed)
	 57 - check-with-native-calc-multiplication-mips (Failed)
	 59 - check-with-qemu-calc-multiplication-mips (Failed)
	 61 - compile-translated-function_call-mips (Failed)
	 62 - run-translated-test-function_call-default-mips (Failed)
	 63 - check-with-native-function_call-default-mips (Failed)
	 65 - check-with-qemu-function_call-default-mips (Failed)
	 67 - compile-translated-floating_point-mips (Failed)
	 68 - run-translated-test-floating_point-default-mips (Failed)
	 69 - check-with-native-floating_point-default-mips (Failed)
	 71 - check-with-qemu-floating_point-default-mips (Failed)
	 73 - compile-translated-syscall-mips (Failed)
	 74 - run-translated-test-syscall-default-mips (Failed)
	 75 - check-with-native-syscall-default-mips (Failed)
	 77 - check-with-qemu-syscall-default-mips (Failed)
	 79 - compile-translated-global-mips (Failed)
	 80 - run-translated-test-global-default-mips (Failed)
	 81 - check-with-native-global-default-mips (Failed)
	 83 - check-with-qemu-global-default-mips (Failed)
	 85 - compile-translated-calc-x86_64 (Failed)
	 86 - run-translated-test-calc-literal-x86_64 (Failed)
	 87 - check-with-native-calc-literal-x86_64 (Failed)
	 89 - check-with-qemu-calc-literal-x86_64 (Failed)
	 90 - run-translated-test-calc-sum-x86_64 (Failed)
	 91 - check-with-native-calc-sum-x86_64 (Failed)
	 93 - check-with-qemu-calc-sum-x86_64 (Failed)
	 94 - run-translated-test-calc-multiplication-x86_64 (Failed)
	 95 - check-with-native-calc-multiplication-x86_64 (Failed)
	 97 - check-with-qemu-calc-multiplication-x86_64 (Failed)
	 99 - compile-translated-function_call-x86_64 (Failed)
	100 - run-translated-test-function_call-default-x86_64 (Failed)
	101 - check-with-native-function_call-default-x86_64 (Failed)
	103 - check-with-qemu-function_call-default-x86_64 (Failed)
	105 - compile-translated-floating_point-x86_64 (Failed)
	106 - run-translated-test-floating_point-default-x86_64 (Failed)
	107 - check-with-native-floating_point-default-x86_64 (Failed)
	109 - check-with-qemu-floating_point-default-x86_64 (Failed)
	111 - compile-translated-syscall-x86_64 (Failed)
	112 - run-translated-test-syscall-default-x86_64 (Failed)
	113 - check-with-native-syscall-default-x86_64 (Failed)
	115 - check-with-qemu-syscall-default-x86_64 (Failed)
	117 - compile-translated-global-x86_64 (Failed)
	118 - run-translated-test-global-default-x86_64 (Failed)
	119 - check-with-native-global-default-x86_64 (Failed)
	121 - check-with-qemu-global-default-x86_64 (Failed)
Errors while running CTest
```

**Comment by aleclearmind** [comment]
Please attach:

* `Testing/Temporary/LastTest.log`
* the binary program, not the source code
* the segfaulating translated program

Tests are failing due to the missing `-no-pie`, I'm going to push some changes that fix this.
The interesting issue is why it segfaults after you add `-no-pie`.

**Comment by aleclearmind** [comment]
Actually, probably you just a value of `/proc/sys/vm/mmap_min_addr` too high.

Running `echo 1024 | sudo tee /proc/sys/vm/mmap_min_addr` should fix the issue.
If you run the current develop branch (abd3154c0341b69a5b47de272d667a13de545b19) you should get a warning about this when running the translate script. The `-no-pie` issue should be now fixed too (55eb769aae6728cf24ba35114e3734d30eaa2781).

**Comment by sdasgup3** [comment]
Fixing the  high value solves the problem. Thanks :)
